### PR TITLE
Fix clipboard on Wayland

### PR DESCRIPTION
**Pull request by Detegr**
arboard advertises that it works with Wayland, but in reality it only works with Wayland terminal applications. To make the clipboard work with applications that draw Wayland surfaces, arboard isn't going to work.

Copypasta does support Wayland's graphical clipboard, but the usage isn't documented. However, for the reasons mentioned in #1474 the move from Copypasta to arboard makes sense.

To resolve the issue, this commit brings in an optional dependency smithay-clipboard, that is a crate that correctly handles the Wayland clipboard in graphical applications. It is used by default if a Wayland window handle is found. If for some reason the handle to the Wayland window handle cannot be fetched, arboard is used as a backup.

It feels a bit wrong adding the `wayland_display` argument to `from_pixels_per_point`, but I couldn't figure out any better alternatives. Maybe pass the whole window to it?

Fixes https://github.com/emilk/egui/issues/1612

### CODE REVIEW

**Comment by emilk** [comment]
wouldn't it make more sense to make this PR to `arboard`?

**Comment by Detegr** [comment]
Maybe. I'll try preparing and submitting a PR to them. Anyway, `egui` needs a modification to pass the wayland display handle to the clipboard crate, whatever it might be, so I'd leave this open for now.

File: egui-winit/Cargo.toml
Comment (emilk) [inline egui-winit/Cargo.toml:52]: We should put this under `[target.'cfg(target_os = "linux")'.dependencies]`

and then add a`#[cfg(target_os = "linux")]` to all uses of it, or
`#[cfg(all(target_os = "linux", feature = "smithay-clipboard"))]`

**Review by emilk — CHANGES_REQUESTED** [review]

**Comment by emilk** [comment]
> Maybe. I'll try preparing and submitting a PR to them. Anyway, `egui` needs a modification to pass the wayland display handle to the clipboard crate, whatever it might be, so I'd leave this open for now.

Let's merge this fix into `egui-winit` now and then, later, we can hopefully remove it when this gets upstreamed to `arboard`.

File: egui-winit/Cargo.toml
Comment (Detegr) [reply egui-winit/Cargo.toml]: I went for `cfg(unix)` everywhere instead as at least FreeBSD has [some kind of support](https://docs.freebsd.org/en/books/handbook/x11/#x-wayland) for Wayland.

File: egui-winit/src/lib.rs
Comment (Detegr) [inline egui-winit/src/lib.rs:64]: I think this needs to be wrapped in conditional compilation too, as Windows won't have the `wayland_display()` function that comes from a platform dependent trait.

File: egui-winit/Cargo.toml
Comment (Detegr) [reply egui-winit/Cargo.toml]: However it seems `winit` is using:
```
#![cfg(any(
    target_os = "linux",
    target_os = "dragonfly",
    target_os = "freebsd",
    target_os = "netbsd",
    target_os = "openbsd"
))]
```
instead for the unix platform module. I guess that means that compiling this as-is will fail on some esoteric operating system that is an unix but isn't any of those in the list.

File: egui-winit/src/lib.rs
Comment (emilk) [reply egui-winit/src/lib.rs:64]: I added Windows and Max to the CI matrix (in https://github.com/emilk/egui/pull/1631), so if you merge in `master` to this PR branch the CI will test for you!

Comment (Detegr) [reply egui-winit/src/lib.rs:64]: Nice! A welcome change. Rebased `master`. I guess I need a maintainer approval to get it the CI running as I'm a first-time contributor.

**Comment by emilk** [comment]
So it seems the new CI checks works! The problem with `#[cfg(not(unix))]` is that it includes macos!

**Comment by enomado** [comment]
cargo check --target wasm32-unknown-unknown

emits

    69 |         Ok(Clipboard { platform: PlatformClipboard::new()? })
       |                                  ^^^^^^^^^^^^^^^^^ use of undeclared type `PlatformClipboard`

**Comment by Detegr** [comment]
I see now why `#[cfg(unix)]` is problematic. Let's use the same OS listing as `winit` does for the conditional compilation, that should work I guess.

**Comment by Detegr** [comment]
> cargo check --target wasm32-unknown-unknown
> 
> emits
> 
> ```
> 69 |         Ok(Clipboard { platform: PlatformClipboard::new()? })
>    |                                  ^^^^^^^^^^^^^^^^^ use of undeclared type `PlatformClipboard`
> ```

I think this issue is unrelated to this PR. The same happens with the `master` branch too.

**Comment by emilk** [comment]
I would prefer if you made new commits instead of squashing. New commits means I can see the changes since the previous one! I always squash on merge anyway, so don't worry about making a pretty history.

**Comment by Detegr** [comment]
You can use the `Compare` button to see changes. But if you're squashing them on merge anyway, I'll do separate commits in the future :)

**Comment by emilk** [comment]
> You can use the `Compare` button to see changes

Today I learned! thanks :)

**Pull request merged**
